go/chat/storage.go

```diff
@@ -279,22 +332,114 @@ func (s *Storage) updateSupersededBy(bi blockIndex, msgs []chat1.MessageFromServ
 		// Update supersededBy on the target message if we have it
 		pos := s.getBlockPosition(superID)
 		superMsg := &b.Msgs[pos]
-		if superMsg.Message != nil {
-			s.debug("updateSupersededBy: writing: id: %d superseded: %d blockID: %d pos: %d",
-				msgid, superID, b.BlockID, pos)
-			superMsg.Message.ServerHeader.SupersededBy = msgid
-			if err = s.writeBlock(bi, b); err != nil {
-				return err
-			}
-		} else {
-			s.debug("updateSupersededBy: skipping id: %d, it is stored as an error", msgid)
+		s.debug("updateSupersededBy: writing: id: %d superseded: %d blockID: %d pos: %d",
+			msgid, superID, b.BlockID, pos)
+		if err = s.updateSupersededBy(ctx, superMsg, uid, msgid); err != nil {
+			return err
+		}
+		if err = s.writeBlock(bi, b); err != nil {
+			return err
 		}
 	}
 
 	return nil
 }
 
-func (s *Storage) writeMessages(bi blockIndex, msgs []chat1.MessageFromServerOrError) libkb.ChatStorageError {
+func (s *Storage) getSecretBoxKey() (fkey [32]byte, err error) {
+
+	// Get secret device key
+	encKey, err := engine.GetMySecretKey(s.G(), s.getSecretUI, libkb.DeviceEncryptionKeyType, "encrypt chat message")
+	if err != nil {
+		return fkey, err
+	}
+	kp, ok := encKey.(libkb.NaclDHKeyPair)
+	if !ok || kp.Private == nil {
+		return fkey, libkb.KeyCannotDecryptError{}
+	}
+
+	// Derive symmetric key from device key
+	skey, err := encKey.SecretSymmetricKey(libkb.EncryptionReasonChatLocalStorage)
+	if err != nil {
+		return fkey, err
+	}
+
+	copy(fkey[:], skey)
+	return fkey, nil
+}
+
+func (s *Storage) encryptMessages(ctx context.Context, uid gregor1.UID,
+	msgs []chat1.MessageFromServerOrError) ([]boxedLocalMessage, libkb.ChatStorageError) {
+
+	s.debug("encrypting %d messages", len(msgs))
+	fkey, err := s.getSecretBoxKey()
+	if err != nil {
+		return nil, libkb.ChatStorageMiscError{Msg: fmt.Sprintf("encryptMessage: failed to get secret key: uid: %s err: %s", uid, err.Error())}
+	}
+
+	var res []boxedLocalMessage
+	for _, msg := range msgs {
+		// Encode message
+		encmsg, err := s.encode(msg)
+		if err != nil {
+			return nil, libkb.ChatStorageMiscError{Msg: fmt.Sprintf("encryptMessage: failed to encode msg: %s", err.Error())}
+		}
+
+		// Encrypt message with key derived from device key
+		var nonce []byte
+		nonce, err = libkb.RandBytes(24)
+		if err != nil {
+			return nil, libkb.ChatStorageMiscError{Msg: fmt.Sprintf("encryptMessage: failure to generate nonce: %s", err.Error())}
+		}
+		var fnonce [24]byte
+		copy(fnonce[:], nonce)
+		sealed := secretbox.Seal(nil, encmsg, &fnonce, &fkey)
+
+		res = append(res, boxedLocalMessage{
+			Version:    1,
+			MsgID:      msg.GetMessageID(),
+			MsgType:    msg.GetMessageType(),
+			Ciphertext: sealed,
+			Nonce:      fnonce,
+		})
+	}
+
+	return res, nil
+}
+
+func (s *Storage) decryptMessages(ctx context.Context, uid gregor1.UID,
+	msgs []boxedLocalMessage) ([]chat1.MessageFromServerOrError, libkb.ChatStorageError) {
+
+	s.debug("decrypting %d messages", len(msgs))
+
+	// Get decryption key
```

Comment:
assert that we're getting version=1 back out here since we don't know how to decrypt 2.